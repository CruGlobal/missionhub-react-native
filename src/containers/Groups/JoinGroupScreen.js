import React, { Component } from 'react';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';

import { Flex, Text } from '../../components/common';
import Header from '../Header';
import BackButton from '../BackButton';

@translate('groupsJoinGroup')
class JoinGroupScreen extends Component {
  render() {
    return (
      <Flex value={1}>
        <Header left={<BackButton />} title="Join Group" />
        <Text>Join Group</Text>
      </Flex>
    );
  }
}

JoinGroupScreen.propTypes = {};

const mapStateToProps = (state, { navigation }) => ({
  ...(navigation.state.params || {}),
});

export default connect(mapStateToProps)(JoinGroupScreen);
export const JOIN_GROUP_SCREEN = 'nav/JOIN_GROUP_SCREEN';
