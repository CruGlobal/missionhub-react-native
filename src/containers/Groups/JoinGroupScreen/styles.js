import { StyleSheet } from 'react-native';

import theme from '../../../theme';

export default StyleSheet.create({
  flex: {
    flex: 1,
  },
  container: {
    flex: 1,
    backgroundColor: theme.primaryColor,
    justifyContent: 'space-between',
    flexDirection: 'column',
  },
  imageWrap: {
    marginTop: 5,
    marginHorizontal: 20,
    minHeight: theme.fullHeight * 0.3,
  },
  text: {
    fontSize: 16,
    color: theme.white,
    textAlign: 'center',
    padding: 4,
  },
  fieldWrap: {
    marginTop: 5,
    paddingTop: 10,
    paddingHorizontal: 30,
  },
  input: {
    fontSize: 36,
    color: theme.white,
    textAlign: 'center',
    padding: 8,
  },
  searchButton: {
    width: theme.fullWidth,
  },
});
