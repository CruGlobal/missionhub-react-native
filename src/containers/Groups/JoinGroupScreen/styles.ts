import { StyleSheet } from 'react-native';

import theme from '../../../theme';

export default StyleSheet.create({
  flex: {
    flex: 1,
  },
  container: {
    flex: 1,
    backgroundColor: theme.primaryColor,
    justifyContent: 'space-between',
    flexDirection: 'column',
  },
  imageWrap: {
    marginTop: 5,
    minHeight: theme.fullHeight * 0.3,
  },
  text: {
    ...theme.textRegular16,
    color: theme.white,
    textAlign: 'center',
    padding: 4,
  },
  fieldWrap: {
    marginTop: 5,
    paddingTop: 10,
    paddingHorizontal: 30,
  },
  input: {
    ...theme.textRegular16,
    fontSize: 36,
    color: theme.white,
    textAlign: 'center',
    padding: 8,
  },
  searchButton: {
    width: theme.fullWidth,
  },
});
