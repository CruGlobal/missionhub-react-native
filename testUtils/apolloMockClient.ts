import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema, IMocks } from 'graphql-tools';
import { DocumentNode } from 'graphql';
import { buildClientSchema, IntrospectionQuery } from 'graphql/utilities';
import gql from 'fraql';
import { createMockerFromIntrospection } from 'fraql/mock';

import introspectionQuery from '../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks: IMocks = {}) => {
  const schema = buildClientSchema(
    (introspectionQuery as unknown) as IntrospectionQuery,
  );

  addMockFunctionsToSchema({
    schema,
    mocks: { ...globalMocks, ...mocks },
  });

  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new SchemaLink({ schema }),
  });
};

export const mockFragment = <TData>(
  fragmentDocument: DocumentNode,
  options?: {
    mocks: IMocks;
  },
): TData => {
  if (!fragmentDocument.loc) {
    throw 'Unable to get original gql query to create FraQL fragment.';
  }

  return createMockerFromIntrospection(introspectionQuery, {
    mocks: globalMocks,
  }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
