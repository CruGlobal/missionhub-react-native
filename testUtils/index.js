import 'react-native';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Need to import for all translation screens to work properly
import '../src/i18n';

Enzyme.configure({ adapter: new Adapter() });

export const createMockStore = (state = {}) => {
  return {
    getState: jest.fn(() => (state)),
    dispatch: jest.fn((response) => Promise.resolve(response)),
    subscribe: jest.fn(),
  };
};

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (data) => {
  expect(renderer.create(data)).toMatchSnapshot();
};

export const renderShallow = (component, store) => {
  let renderedComponent = shallow(
    component,
    { context: { store: store } }
  );

  // If component has translation wrappers, dive deeper
  while (renderedComponent.is('Translate') || renderedComponent.is('I18n')) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (component, store) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};

export const mockFnWithParams = (obj, method, expectedReturn, ...expectedParams) => {
  return obj[method] = jest.fn(
    (...actualParams) => {
      expect(actualParams).toEqual(expectedParams);
      return expectedReturn;
    }
  );
};
