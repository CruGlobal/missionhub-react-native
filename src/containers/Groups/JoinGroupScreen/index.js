import React, { Component } from 'react';
import {
  ScrollView,
  View,
  Keyboard,
  Image,
  KeyboardAvoidingView,
} from 'react-native';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';

import {
  Flex,
  Text,
  Input,
  IconButton,
  Button,
} from '../../../components/common';
import Header from '../../Header';
import theme from '../../../theme';
import GROUP_ICON from '../../../../assets/images/MemberContacts_light.png';
import { navigateBack } from '../../../actions/navigation';
import ImagePicker from '../../../components/ImagePicker';

import styles from './styles';

@translate('groupsJoinGroup')
class JoinGroupScreen extends Component {
  state = {
    code: '',
    imageUri: null,
  };

  onChangeCode = code => this.setState({ code });

  onSearch = () => {
    Keyboard.dismiss();
    const text = (this.state.name || '').trim();
    if (!text) {
      return;
    }

    // TODO: search community by code
  };

  joinCommunity = () => {
    Keyboard.dismiss();
    // TODO: join community
  };

  handleImageChange = data => this.setState({ imageUri: data.uri });

  navigateBack = () => this.props.dispatch(navigateBack());

  ref = c => (this.nameInput = c);

  renderImage() {
    const { imageUri } = this.state;
    if (imageUri) {
      return (
        <Image
          resizeMode="cover"
          source={{ uri: imageUri }}
          style={styles.image}
        />
      );
    }
    return <Image source={CAMERA_ICON} />;
  }

  renderStart() {
    const { t } = this.props;
    return (
      <Flex align="center" justify="center">
        <Image resizeMode="contain" source={GROUP_ICON} style={styles.image} />
        <Text style={styles.text}>{t('enterCode')}</Text>
      </Flex>
    );
  }

  renderError() {
    const { t } = this.props;
    return (
      <Flex align="center" justify="center">
        <Text style={styles.text}>{t('communityNotFound')}</Text>
      </Flex>
    );
  }

  render() {
    const { t } = this.props;

    return (
      <View style={styles.container}>
        <Header
          left={
            <IconButton
              name="deleteIcon"
              type="MissionHub"
              onPress={this.navigateBack}
            />
          }
          shadow={false}
          title={t('joinCommunity')}
        />
        <ScrollView keyboardShouldPersistTaps="handled" style={styles.flex}>
          <Flex align="center" justify="end" style={styles.imageWrap}>
            {this.renderStart()}
          </Flex>
          <Flex style={styles.fieldWrap}>
            <Input
              style={styles.input}
              onChangeText={this.onChangeCode}
              maxLength={6}
            />
          </Flex>
          <KeyboardAvoidingView
            keyboardVerticalOffset={theme.buttonHeight}
            style={styles.flex}
          />
        </ScrollView>

        <Flex align="stretch" justify="end">
          <Button
            type="secondary"
            onPress={this.onSearch}
            text={t('search').toUpperCase()}
            style={styles.searchButton}
          />
        </Flex>
      </View>
    );
  }
}

JoinGroupScreen.propTypes = {};

const mapStateToProps = (state, { navigation }) => ({
  ...(navigation.state.params || {}),
});

export default connect(mapStateToProps)(JoinGroupScreen);
export const JOIN_GROUP_SCREEN = 'nav/JOIN_GROUP_SCREEN';
