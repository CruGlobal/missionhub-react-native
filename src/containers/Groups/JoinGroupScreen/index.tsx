import React, { Component } from 'react';
import {
  ScrollView,
  Keyboard,
  Image,
  KeyboardAvoidingView,
  View,
  Text,
} from 'react-native';
import { connect } from 'react-redux-legacy';
import { withTranslation } from 'react-i18next';
import PropTypes from 'prop-types';

import { Flex, Input } from '../../../components/common';
import GroupCardItem from '../../../components/GroupCardItem';
import Header from '../../../components/Header';
import theme from '../../../theme';
import GROUP_ICON from '../../../../assets/images/MemberContacts_light.png';
import { lookupOrgCommunityCode } from '../../../actions/organizations';
import CloseButton from '../../../components/CloseButton';
import Analytics from '../../Analytics';

import styles from './styles';

// @ts-ignore
@withTranslation('groupsJoinGroup')
class JoinGroupScreen extends Component {
  state = {
    code: '',
    errorMessage: '',
    community: undefined,
  };

  componentDidMount() {
    // Pause before focus so that the other screen can disappear before the keyboard comes up
    setTimeout(
      // @ts-ignore
      () => this.codeInput && this.codeInput.focus && this.codeInput.focus(),
      350,
    );
  }

  // @ts-ignore
  onChangeCode = code => {
    this.setState({ code: code.toUpperCase() }, () => {
      if (code.length >= 6) {
        this.onSearch();
      }
    });
  };

  onSearch = async () => {
    const {
      // @ts-ignore
      codeInput,
      state: { code },
      // @ts-ignore
      props: { t, dispatch },
    } = this;

    codeInput.focus();

    const errorState = {
      errorMessage: t('communityNotFound'),
      community: undefined,
    };

    const text = (code || '').trim();
    if (!text || text.length < 6) {
      this.setState(errorState);
      return;
    }

    try {
      const community = await dispatch(lookupOrgCommunityCode(text));

      if (!community) {
        this.setState(errorState);
        return;
      }
      this.setState({ errorMessage: '', community });
    } catch (e) {
      this.setState(errorState);
    }
  };

  navigateNext = () => {
    // @ts-ignore
    const { dispatch, next } = this.props;
    const { community } = this.state;
    Keyboard.dismiss();

    dispatch(
      next({
        community,
      }),
    );
  };

  // @ts-ignore
  ref = c => (this.codeInput = c);

  renderStart() {
    // @ts-ignore
    const { t } = this.props;
    return (
      <Flex align="center" justify="center">
        {/* 
        // @ts-ignore */}
        <Image resizeMode="contain" source={GROUP_ICON} style={styles.image} />
        <Text style={styles.text}>{t('enterCode')}</Text>
      </Flex>
    );
  }

  renderError() {
    const { errorMessage } = this.state;
    return (
      <Flex align="center" justify="center">
        <Text style={styles.text}>{errorMessage}</Text>
      </Flex>
    );
  }

  renderGroupCard() {
    const { community } = this.state;

    const {
      // @ts-ignore
      id,
      // @ts-ignore
      name,
      // @ts-ignore
      owner,
      // @ts-ignore
      contactReport = {},
      // @ts-ignore
      community_photo_url,
      // @ts-ignore
      unread_comments_count,
    } = community;

    const group = {
      id,
      name,
      owner: {
        nodes: owner
          ? [
              {
                firstName: owner.first_name,
                lastName: owner.last_name,
              },
            ]
          : [],
      },
      report: {
        contactCount: contactReport.contactsCount || 0,
        unassignedCount: contactReport.unassignedCount || 0,
        memberCount: contactReport.memberCount || 0,
      },
      communityPhotoUrl: community_photo_url,
      unreadCommentsCount: unread_comments_count,
    };

    // @ts-ignore
    return <GroupCardItem group={group} onJoin={this.navigateNext} />;
  }

  render() {
    // @ts-ignore
    const { t } = this.props;
    const { code, errorMessage, community } = this.state;

    return (
      <View style={styles.container}>
        <Analytics screenName={['communities', 'enter code']} />
        <Header left={<CloseButton />} title={t('joinCommunity')} />
        <ScrollView keyboardShouldPersistTaps="handled" style={styles.flex}>
          <Flex align="center" justify="end" style={styles.imageWrap}>
            {errorMessage
              ? this.renderError()
              : community
              ? this.renderGroupCard()
              : this.renderStart()}
          </Flex>
          <Flex style={styles.fieldWrap}>
            <Input
              ref={this.ref}
              style={styles.input}
              onChangeText={this.onChangeCode}
              selectionColor={theme.white}
              maxLength={6}
              value={code}
              autoFocus={false}
              blurOnSubmit={false}
              testID="joinInput"
            />
          </Flex>
          <KeyboardAvoidingView
            keyboardVerticalOffset={theme.buttonHeight}
            style={styles.flex}
          />
        </ScrollView>
      </View>
    );
  }
}

// @ts-ignore
JoinGroupScreen.propTypes = {
  next: PropTypes.func.isRequired,
};

export default connect()(JoinGroupScreen);
export const JOIN_GROUP_SCREEN = 'nav/JOIN_GROUP_SCREEN';
