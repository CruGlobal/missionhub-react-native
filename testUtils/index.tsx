import React, { ReactElement, ReactNode } from 'react';
import 'react-native';
import { Provider } from 'react-redux';
import { Provider as ProviderLegacy } from 'react-redux-legacy';
import thunk from 'redux-thunk';
import configureStore, { MockStore } from 'redux-mock-store';
// eslint-disable-next-line import/named
import { NavigationParams, NavigationProvider } from 'react-navigation';
import { ApolloProvider } from '@apollo/react-hooks';
import { ReactTestRendererJSON } from 'react-test-renderer';
import { render, RenderAPI } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';
import Enzyme, { shallow as enzymeShallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { IMocks } from 'graphql-tools';

import { createApolloMockClient } from './apolloMockClient';
import { createNavigationProp } from './navigationHelpers';

Enzyme.configure({ adapter: new Adapter() });

export const createThunkStore = configureStore([thunk]);

interface RenderWithContextParams {
  initialState?: {};
  store?: MockStore;
  navParams?: NavigationParams;
  mocks?: IMocks;
  noWrappers?: boolean;
}

// Inspiration from https://github.com/kentcdodds/react-testing-library/blob/52575005579307bcfbe7fbe4ef4636147c03c6fb/examples/__tests__/react-redux.js#L69-L80
export function renderWithContext(
  component: ReactElement,
  {
    initialState,
    store = createThunkStore(initialState),
    navParams,
    mocks: mocks = {},
    noWrappers = false,
  }: RenderWithContextParams = {},
) {
  const mockApolloClient = createApolloMockClient(mocks);

  const navigation = createNavigationProp(navParams);

  let renderResult: RenderAPI;
  if (noWrappers) {
    renderResult = render(React.cloneElement(component, { navigation }));
  } else {
    // Warning: don't call any functions in here that return new instances on every call. All the props need to stay the same otherwise rerender won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
      <NavigationProvider value={navigation}>
        <ProviderLegacy store={store}>
          <Provider store={store}>
            <ApolloProvider client={mockApolloClient}>
              {children}
            </ApolloProvider>
          </Provider>
        </ProviderLegacy>
      </NavigationProvider>
    );

    renderResult = render(React.cloneElement(component, { navigation }), {
      wrapper,
    });
  }

  let storedSnapshot: ReactTestRendererJSON | null;
  return {
    ...renderResult,
    store,
    snapshot: () => expect(renderResult.toJSON()).toMatchSnapshot(),
    recordSnapshot: () => (storedSnapshot = renderResult.toJSON()),
    diffSnapshot: () => {
      if (!storedSnapshot) {
        throw new Error(
          'You must call recordSnapshot to store an initial snapshot before calling diffSnapshot',
        );
      }
      expect(
        snapshotDiff(storedSnapshot, renderResult.toJSON()),
      ).toMatchSnapshot();
    },
  };
}

// TODO: Remove all legacy rendering functions below

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (component: ReactElement) => {
  const { toJSON } = render(component);
  expect(toJSON()).toMatchSnapshot();
};

export const renderShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  let renderedComponent = enzymeShallow(
    <ProviderLegacy store={store}>{component}</ProviderLegacy>,
  ).dive();

  // If component has translation wrappers, dive deeper
  while (
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    ((renderedComponent.type() as any).displayName || '').startsWith(
      'withI18nextTranslation(',
    )
  ) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};
