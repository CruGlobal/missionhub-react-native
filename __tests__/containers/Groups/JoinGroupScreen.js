import React from 'react';
import configureStore from 'redux-mock-store';

import JoinGroupScreen from '../../../src/containers/Groups/JoinGroupScreen';
import { renderShallow, createMockNavState } from '../../../testUtils';

const mockStore = configureStore();
const organizations = {
  all: [
    {
      id: '1',
      name: 'Test Org 1',
      contactReport: {},
    },
    {
      id: '2',
      name: 'Test Org 2',
      contactReport: {},
      user_created: true,
    },
  ],
};
const auth = {};
const store = mockStore({ organizations, auth });

it('should render null state', () => {
  const component = renderShallow(
    <JoinGroupScreen navigation={createMockNavState()} />,
    mockStore({ organizations: { all: [] }, auth }),
  );
  expect(component).toMatchSnapshot();
});

describe('JoinGroupScreen', () => {
  let component;
  beforeEach(() => {
    component = renderShallow(
      <JoinGroupScreen navigation={createMockNavState()} />,
      store,
    );
  });

  it('should render correctly', () => {
    expect(component).toMatchSnapshot();
  });
});
