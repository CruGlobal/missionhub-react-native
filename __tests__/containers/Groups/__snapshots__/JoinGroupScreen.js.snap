// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinGroupScreen should render correctly 1`] = `
<Flex
  value={1}
>
  <Connect(Header)
    left={<Connect(BackButton) />}
    title="Join Group"
  />
  <MyText
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Join Group
  </MyText>
</Flex>
`;

exports[`should render null state 1`] = `
<Flex
  value={1}
>
  <Connect(Header)
    left={<Connect(BackButton) />}
    title="Join Group"
  />
  <MyText
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Join Group
  </MyText>
</Flex>
`;
